### ISSUE TITLE
SQLite ANR

### ISSUE THREAD

**Issue by ericntd** [mParticle/mparticle-android-sdk#62]
at sun.misc.Unsafe.park (Unsafe.java)
  at java.util.concurrent.locks.LockSupport.parkNanos (LockSupport.java:230)
  at android.database.sqlite.SQLiteConnectionPool.waitForConnection (SQLiteConnectionPool.java:1031)
  at android.database.sqlite.SQLiteConnectionPool.acquireConnection (SQLiteConnectionPool.java:609)
  at android.database.sqlite.SQLiteSession.acquireConnection (SQLiteSession.java:916)
  at android.database.sqlite.SQLiteSession.prepare (SQLiteSession.java:588)
  at android.database.sqlite.SQLiteProgram.<init> (SQLiteProgram.java:62)
  at android.database.sqlite.SQLiteQuery.<init> (SQLiteQuery.java:37)
  at android.database.sqlite.SQLiteDirectCursorDriver.query (SQLiteDirectCursorDriver.java:46)
  at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory (SQLiteDatabase.java:2063)
  at android.database.sqlite.SQLiteDatabase.queryWithFactory (SQLiteDatabase.java:1910)
  at android.database.sqlite.SQLiteDatabase.query (SQLiteDatabase.java:1781)
  at android.database.sqlite.SQLiteDatabase.query (SQLiteDatabase.java:1949)
  at com.mparticle.internal.a.b.a (b.java:66)
  at com.mparticle.internal.a.a.h.a (h.java:38)
  at com.mparticle.internal.a.a.b.a (b.java:510)
  at com.mparticle.internal.a.a.b.a (b.java:561)
  at com.mparticle.internal.u.a (u.java:853)
  at com.mparticle.identity.c.a (c.java:42)
  at com.mparticle.identity.d.getUserAttributes (d.java:42)
  at com.mparticle.kits.FilteredMParticleUser.getUserAttributes (FilteredMParticleUser.java:51)
  at com.mparticle.kits.LeanplumKit.onUserIdentified (LeanplumKit.java:113)
  at com.mparticle.kits.KitManagerImpl.onUserIdentified (KitManagerImpl.java:1124)
  at com.mparticle.identity.IdentityApi$b$a.run (IdentityApi.java:452)
  at android.os.Handler.handleCallback (Handler.java:938)
  at android.os.Handler.dispatchMessage (Handler.java:99)
  at android.os.Looper.loop (Looper.java:246)
  at android.app.ActivityThread.main (ActivityThread.java:8425)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:596)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1130)

Samsung Galaxy S10
Android 11 (SDK 30)

mParticle library version: 
def mparticle = [
        appsflyer: "com.mparticle:android-appsflyer-kit:5.14.8",
        leanplum : "com.mparticle:android-leanplum-kit:5.14.8",
]

**Comment by willpassidomo** [comment]
hey @ericntd thanks for reporting this! I don't like the looks of this, thread-locks are never good. If I could gather a little more information:

1) how often are you seeing this? If often when, on startup?
2) since this was in a background thread, did it propagate as a Logcat log or did it fully crash the app?

**Comment by ericntd** [comment]
Hi @willpassidomo ,
Unfortunately, i don't have simple and straight answers to your questions
> 1. how often are you seeing this? If often when, on startup?
it's the number one ANR in our app last 30 days
It's happening on a particular Activity
<img width="923" alt="Screen Shot 2021-03-19 at 4 44 53 PM" src="https://user-images.githubusercontent.com/19919208/111840507-e0259080-88d2-11eb-9e70-2cd679bdd0c0.png">

> 2. since this was in a background thread, did it propagate as a Logcat log or did it fully crash the app?
This is an ANR, not a crash
Unfortunately, our small team have not reproduced it in-house :(

**Comment by willpassidomo** [comment]
No problem, that is very useful information. I think it's very interesting that it is happening in one place in particular, would you be able to share your project, or if not, some of the relevant mparticle usage leading up to where the ANR occurs?

**Comment by ericntd** [comment]
Hi @willpassidomo 

Our mParticle logics can be found below
A question though:
- Do `login()` and co. need to happen on main thread, or can they run in background? If yes, I'd move them to background and the ANR should be go away. Wdyt?

```java
public class MParticleManager implements MParticleContract.Manager {

  private static final String TAG = "MParticleManager";
  private static final int MAX_RETRY_ATTEMPTS = 10;
  private static final long RETRY_INTERVAL = 5000;

  private final FacebookContract.Manager facebookManager;
  private int retryAttempts = 0;

  public MParticleManager(
      FacebookContract.Manager facebookManager) {
    this.facebookManager = facebookManager;
  }

  @Override
  public void login() {
    MParticle.getInstance().Identity().login(IdentityApiRequest.withEmptyUser().build())
             .addFailureListener(response -> {
               if (isRetryableResponseCode(response.getHttpCode())) {
                 retryMParticleAction(MParticleManager.this::login, response);
               } else {
                 logFailure(response);
               }
             })
             .addSuccessListener(result -> retryAttempts = 0);
  }

  @Override
  public void logout() {
    MParticle.getInstance().Identity().logout(IdentityApiRequest.withEmptyUser().build())
             .addFailureListener(this::logFailure)
             .addSuccessListener(result -> retryAttempts = 0);
  }

  @Override
  public void logEvent(
      String eventName, MParticle.EventType eventType,
      Map<String, String> attributes) {
    MPEvent event = new MPEvent.Builder(eventName, eventType).info(attributes).build();
    MParticle.getInstance().logEvent(event);
  }

  @Override
  public void setUserAttribute(String key, Object value) {
    MParticleUser currentUser = MParticle.getInstance().Identity().getCurrentUser();
    if (currentUser != null) {
      currentUser.setUserAttribute(key, value);
    }
  }

  @Override
  public void logScreen(String screen) {
    MParticle.getInstance().logScreen(screen);
    MParticle.getInstance().upload();
  }

  @Override
  public void modifyIdentity(Profile myOwnProfile, Subscription subscription) {
    MParticle.getInstance().Identity().modify(getApiRequest(myOwnProfile))
             .addFailureListener(response -> {
               if (isRetryableResponseCode(response.getHttpCode())) {
                 retryMParticleAction(() -> modifyIdentity(myOwnProfile, subscription), response);
               } else {
                 logFailure(response);
               }
             })
             .addSuccessListener(result -> retryAttempts = 0);
  }

  @Override
  public void modifyIdentity(String profileId) {
    MParticle.getInstance().Identity().modify(
        IdentityApiRequest.withEmptyUser().customerId(profileId).build())
             .addFailureListener(this::logFailure)
             .addSuccessListener(result -> {
             });
  }

  private IdentityApiRequest getApiRequest(Profile myOwnProfile) {
    return IdentityApiRequest.withEmptyUser()
                             .customerId(myOwnProfile.getId())
                             .email(myOwnProfile.getUserEmail())
                             .userIdentity(
                                 MParticle.IdentityType.Facebook,
                                 facebookManager.getFacebookId())
                             .build();
  }

  private void retryMParticleAction(Runnable runnable, IdentityHttpResponse response) {
    if (retryAttempts <= MAX_RETRY_ATTEMPTS) {
      retryAttempts++;

      Logger.d(TAG, "Retrying mParticle action, attempt " + retryAttempts);

      Handler handler = new Handler();
      handler.postDelayed(runnable, RETRY_INTERVAL);
    } else {
      logFailure(response);
    }
  }

  // Unknown error is thrown when a user doesn't have service.
  // Bad request can occur on the login call due to the login call needing an Identity Type, which in our case is the profile id
  // and the profile object has not been created yet
  private boolean isRetryableResponseCode(int responseCode) {
    return responseCode == IdentityApi.BAD_REQUEST || responseCode == IdentityApi.UNKNOWN_ERROR
        || responseCode == IdentityApi.THROTTLE_ERROR;
  }

  private void logFailure(IdentityHttpResponse response) {
      Logger.d(TAG, "Failed mParticle action " + response.toString());
      Logger.logException(new Throwable("Failed mParticle action " + response.toString()));
  }
}
```

**Comment by willpassidomo** [comment]
So, the login network request is happening in a background thread, but we trigger the onSuccess/onFailure/onUserIdentified callbacks on the main thread. The stack trace you shared originates with an internal `onUserIdentified()` callback registered by the LeanplumKit. 

The problem I see is that we are reading from the database on the main thread. `MParticleUser.getUserAttributes()` is our *only* API that directly accesses the database from the calling thread, but there is an async alternative, `MParticleUser.getUserAttributes(UserAttributeListener)`.

I think the fix here is for us to update the LeanplumKit so that we are using the async method, `MParticleFilteredUser.getUserAttributes(UserAttributeListener)`

**Comment by ericntd** [comment]
@willpassidomo , that makes sense.

Are you going release a fix or is there any action required on our end?

**Comment by willpassidomo** [comment]
@ericntd I've put up a pull request for this [here](https://github.com/mparticle-integrations/mparticle-android-integration-leanplum/pull/49). Barring any unforeseen issues, it will go out in next week's release. Probably too late to sneak it into this week's

When it is released, all you will need to do is upgrade your `android-leanplum-kit` dependency to the latest version

**Comment by willpassidomo** [comment]
Hey @ericntd I'm not sure if you/ve seen it yet or not, but the fix for this was released in [version 5.16.5](https://github.com/mParticle/mparticle-android-sdk/releases/tag/v5.16.5)
